polly/src/event_manager.rs

```diff
@@ -0,0 +1,343 @@
+        None
+    }
+
+    fn handle_write(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+
+    fn handle_close(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+
+    /// Initial registration of pollable objects.
+    fn init(&self) -> Option<Vec<PollableOp>>;
+}
+
+struct HandlerMap {
+    handlers: HashMap<i32, EventHandlerData>,
+}
+
+impl HandlerMap {
+    pub fn new() -> HandlerMap {
+        HandlerMap {
+            handlers: HashMap::new(),
+        }
+    }
+
+    pub fn insert(&mut self, id: i32, handler_data: EventHandlerData) {
+        self.handlers.insert(id, handler_data);
+    }
+
+    pub fn remove(&mut self, id: &i32) -> Option<EventHandlerData> {
+        self.handlers.remove(id)
+    }
+
+    // Returns a copy instead of ref
+    pub fn get(&mut self, id: &i32) -> Option<EventHandlerData> {
+        match self.handlers.get(id) {
+            Some(handler) => Some(EventHandlerData::new(
+                (handler.data.0.clone(), handler.data.1),
+                handler.handler.clone(),
+            )),
+            None => None,
+        }
+    }
+}
+
+impl Deref for HandlerMap {
+    type Target = HashMap<i32, EventHandlerData>;
+    fn deref(&self) -> &Self::Target {
+        &self.handlers
+    }
+}
+
+pub struct EventManager {
+    fd: Pollable,
+    handlers: HandlerMap,
+    events: Vec<epoll::Event>,
+}
+
+impl EventManager {
+    pub fn new() -> Result<EventManager> {
+        let epoll_fd = OwnedFD::from_unowned(epoll::create(true).map_err(|_| Error::EpollCreate)?);
+        Ok(EventManager {
+            fd: epoll_fd,
+            handlers: HandlerMap::new(),
+            events: vec![epoll::Event::new(epoll::Events::empty(), 0); EVENT_BUFFER_SIZE],
+        })
+    }
+
+    fn event_type_to_epoll_mask(event_types: EventType) -> epoll::Events {
+        let mut epoll_event_mask = epoll::Events::empty();
+
+        if event_types.readable() {
+            epoll_event_mask |= epoll::Events::EPOLLIN;
+        }
+
+        if event_types.writeable() {
+            epoll_event_mask |= epoll::Events::EPOLLOUT;
+        }
+
+        if event_types.closed() {
+            epoll_event_mask |= epoll::Events::EPOLLRDHUP;
+        }
+
+        epoll_event_mask
+    }
+
+    fn register_handler(&mut self, event_data: EventRegistrationData, wrapped_handler: Weak<RefCell<dyn EventHandler>>) -> Result<()> {
+        let (pollable, event_type) = event_data;
+
+        if let Some(_) = self.handlers.get(&pollable.clone()) {
+            println!("Pollable {} already registered", pollable.clone());
+            return Err(Error::AlreadyExists);
+        };
+
+        epoll::ctl(
+            self.fd.as_raw_fd(),
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            pollable.as_raw_fd(),
+            epoll::Event::new(
+                EventManager::event_type_to_epoll_mask(event_type),
+                **pollable as u64,
```

Comment:
If you change `Pollable` type to be:
```
type Pollable = Rc<dyn AsRawFd>;
```

then this becomes:
```suggestion
                pollable.as_raw_fd() as u64,
```